Updating Operations API to support Python's AutoRest

sonathan opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

sonathan committed:
Updating Operations API to support Python's AutoRest

msftclas commented:
@sonathan,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

sonathan committed:
Updating readme.md

sonathan committed:
Adding a surpressing message in readme

veronicagg reviewed (changes requested):
Please review model validation errors in added example at https://travis-ci.org/Azure/azure-rest-api-specs/jobs/255998310#L661.
I understand you're not making changes in the rest of the spec, so just making sure you know about linter error/warnings being reported: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/255998308#L674. 
Thanks!

sonathan commented on specification/recoveryservices/resource-manager/readme.md:
````diff
@@ -74,3 +74,24 @@ csharp:
+``` yaml
+directive:
+  - suppress: DefinitionsPropertiesNamesCamelCase
+    reason: Autorest invalidates two letter acronyms as well and changes in data contracts require service wide changes and require more time
````
Yes, we have requested approval for this supprestion already and also for the XmsExamplesProvidedValidation, as said by @ravbhatnagar that "As Samer said, we cant give an exception for the examples violation. This will not block the PR to be merged but it will be shown as a violation till it is fixed." we had a mail thread with azure-spc-violations@microsoft.com, adxsr@microsoft.com with subject: Violations Suppression requested for PR 1139

veronicagg replied:
@sonathan thanks for pointing this out. Regarding the specific violation DefinitionsPropertiesNamesCamelCase the justification in the email and the above doesn't seem to match the reason why the error is being reported. I see properties named as as "Provider" for "ClientDiscoveryDisplay", which violates the rule. I understand this is a breaking change and you may not want to take it right now, would this be a permanent exception? @ravbhatnagar @salameer

ravbhatnagar replied:
@sonathan - We cant add suppression for the camel casing violation. Till this is fixed, it will continue to show as violation.

sonathan commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:
```diff
@@ -50,7 +50,7 @@
         }
       }
     },
-    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/operations": {
```
Thankyou. Removing SubsId and ResourceGroupName from parameter from subsequent commit.

veronicagg replied:
thanks, will take a look in your next commit.

sonathan committed:
removing subsId and RGName from parameters

sonathan commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:
```diff
@@ -50,7 +50,7 @@
       }
     },
-    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/operations": {
+    "/providers/Microsoft.RecoveryServices/operations": {
```
For this you can refer: https://github.com/Azure/azure-rest-api-specs/pull/1267 CLI is not yet released and it will be including these changes.

ravbhatnagar replied:
@veronicagg - This fix is fine since otherwise the API had no meaning and clients couldnt anyway meaningfully use it. And as @sonathan pointed out, the CLI is not yet released so from that perspective it should be ok as well.

ravbhatnagar commented:
@veronicagg - As stated above, lets not add suppression for the camel case fix. It will remain a violation as the service team is expected to fix it.

veronicagg commented:
thanks @ravbhatnagar .
@sonathan please remove suppression that was added to the readme file, then I can approve this PR. thanks!

sonathan committed:
Removing supression from ReadMe.md and updating Examples/ListOperations.json with correct API-Version

veronicagg commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/examples/ListOperations.json:
```diff
@@ -0,0 +1,1339 @@
+                            "Resource": "Vaults",
+                            "Operation": "Resolvealert",
+                            "Description": "Resolves the alert."
+                        },
```
There is a remaining issue with the example model validation https://travis-ci.org/Azure/azure-rest-api-specs/jobs/259348817#L657 
The value for blobDuration doesn't match the date-time format per https://xml2rfc.tools.ietf.org/public/rfc/html/rfc3339.html#anchor14
Please update these.

sonathan replied:
Updated. Thanks Veronica!

veronicagg commented:
@sonathan just let me know if you have questions about the issue pointed out above.

sonathan committed:
Updating examples in Operations API

sonathan commented:
@veronicagg I have updated the examples.

veronicagg commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:
```diff
@@ -456,7 +456,6 @@
         },
         "blobDuration": {
           "description": "Blob duration.",
-          "format": "date-time",
```
is the field really not a date-time?

veronicagg commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/examples/ListOperations.json:
```diff
@@ -1187,42 +1187,6 @@
                         "Origin": "user"
                     },
                     {
-                        "Name": "Microsoft.RecoveryServices/Vaults/providers/Microsoft.Insights/logDefinitions/read",
```
why did you decide to remove this part? It's good to include more and cover more fields, we need to make sure the wire matches what's in the swagger, that's the purpose of having the example covering more.

veronicagg reviewed (changes requested):
questions inline

ravbhatnagar commented:
@veronicagg - I see the label indicating WaitForARMFeedback. Just wanted to confirm if anything is needed from my side on this.

veronicagg commented:
@ravbhatnagar no, you've commented on the suppression not being allowed above, and  the label did not get removed then. I'll remove it now, next time, feel free to remove it after you provide your feedback. thanks!
On this PR I'm waiting on @sonathan regarding my questions on the examples.

sonathan committed:
Updating Client operations API with new contract

sonathan commented:
Hi @veronicagg 

As we are in preview, and the contracts had changed in service side, I have updated those here and have also updated examples to cover out as much scenarios as possible.

veronicagg commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:
```diff
@@ -419,90 +419,78 @@
-          "description": "Name of the log.",
-          "type": "string",
-          "readOnly": true
+          "description": "Name",
```
Please review: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/262193673#L1444

veronicagg commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:
```diff
@@ -419,90 +419,78 @@
+          }
         },
         "NextLink": {
           "description": "NextLink",
```
Please review: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/262193673#L1456

sonathan replied:
Thanks for pointing out.

veronicagg commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:
```diff
@@ -511,25 +499,20 @@
-          "description": "Name of the operation.",
-          "type": "string",
-          "readOnly": true
+          "description": "Name",
```
Please review: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/262193673#L1469

sonathan replied:
Thanks!

veronicagg replied:
@sonathan it appears that this one is still showing up https://travis-ci.org/Azure/azure-rest-api-specs/jobs/262537716#L1457 therefore causing 1 more error in the Linter compared to the spec before this PR, you may want to update this one as well.

veronicagg commented:
@sonathan one more question, since you mentioned you're in preview...
1) are the changes included in this spec deployed?
2) is there a reason why your api-version is not updated or includes "preview" in it?

sonathan commented:
@veronicagg  Yes the changes are already deployed. To be honest we have added this API only for compliance, whenever there is a need we refer to recoveryServicesBackup for this API.

sonathan commented on specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:
```diff
@@ -419,89 +419,78 @@
-          "description": "Blob duration.",
-          "type": "string",
-          "readOnly": true
+          "description": "blob duration",
```
Hi @veronicagg , As this field is in Response and Shoebox creates blobs in customer storage account every hour. The value PT1H indicates blobs per hour. Thanks!

veronicagg replied:
Thanks @sonathan . Should this be part of its description? Are customers expected to know what it means, if so, how? thanks!

sonathan committed:
Correcting error DescriptionMustNotBeNodeName

veronicagg reviewed (changes requested):
Looks good, just one question left and one linter error that appears to not have been updated in the last commit.

veronicagg commented:
@sonathan if you have any questions about my last comment, please let me know, thanks.

veronicagg commented:
@sonathan any updates?

veronicagg commented:
@sonathan FYI - the PR will get closed in a few days if there's no activity from your side. It it gets closed, you may reopen it when you're back to it. Thanks!

veronicagg commented:
@sonathan This PR will get closed tomorrow, if no updates are received, thanks.

veronicagg commented:
@sonathan please reopen the PR and tag me when you're back to this, thanks.
